lib/pbench/test/unit/server/test_unpack_tarballs.py

```diff
@@ -0,0 +1,293 @@
+
+        def mock_resolve(self: Path, strict: bool = False):
+            """
+            Mock replacement for Path.resolve()
+
+            Args:
+                strict: Boolean value to make sure path exist if True.
+
+            Returns:
+                Raising FileNotFoundError Exception
+            """
+            assert strict
+            raise FileNotFoundError(f"No such file or directory: '{self}'")
+
+        monkeypatch.setattr(Path, "glob", lambda path, args: self.tarlist)
+        monkeypatch.setattr(Path, "resolve", mock_resolve)
+        monkeypatch.setattr(Path, "stat", self.mock_stat)
+        obj = UnpackTarballs(server_config, make_logger)
+        result = obj.unpack_tarballs(min_size, max_size)
+        if result.total > 0:
+            for tar in self.tarlist:
+                assert (
+                    f"Tarball link, '{tar}', does not resolve to a real location: No such file or directory: '{tar}'"
+                    in caplog.text
+                )
+        assert result.total == total
+        assert result.success == 0
+
+    def test_unpack_md5_error(self, monkeypatch, server_config, make_logger, caplog):
+        """Show that when md5 file path cannot be accessed it throws a
+        FileNotFoundError Exception"""
+
+        def mock_md5sum(filepath: Path):
+            """
+            Mock of md5sum function
+
+            Args:
+                filepath: Tarball file path
+
+            Returns:
+                Raises FileNotFoundError Exception
+            """
+            raise FileNotFoundError(f"No such file or directory: '{filepath}.md5'")
+
+        monkeypatch.setattr("pbench.server.utils.md5sum", mock_md5sum)
+        file_tree = FileTree(server_config, make_logger)
+        obj = UnpackTarballs(server_config, make_logger)
+        for tar in self.tarlist:
+            with pytest.raises(Exception):
+                obj.unpack(Path(tar), file_tree)
+                assert (
+                    f"Getting md5 value of tarball '{tar}' failed: No such file or directory: '{tar}.md5'"
+                    in caplog.text
+                )
+
+    def test_unpack_exception(self, monkeypatch, server_config, make_logger, caplog):
+        """Show that when unpack module raises TarballUnpackError exception
+        it is being handled properly."""
+
+        def mock_unpack(self, dataset_id):
+            """
+            Mock of FileTree.unpack(self.min_size, self.max_size) function
+
+            Args:
+                dataset_id: tarball md5 value
+
+            Returns:
+                Raises TarballUnpackError Exception
+            """
+            raise TarballUnpackError("tarball", "tar exited with status 1")
+
+        monkeypatch.setattr(
+            "pbench.server.unpack_tarballs.get_tarball_md5", lambda tb: "my_md5_string"
+        )
+        monkeypatch.setattr(FileTree, "unpack", mock_unpack)
+        file_tree = FileTree(server_config, make_logger)
```

Comment:
If you tighten the the test code to explicitly expect `TarballUnpackError` instead of `Exception`, then I think you'll see why the mock is in fact required.

When you pass `None` for the `FileTree` instance as the second argument to `UpackTarballs.unpack()`, then, when the CUT tries to invoke the `unpack()` method on what it expects to be a `FileTree` instance, it will incur an `AttributeError` exception.  While that (mostly) drives the test behavior that you're seeking, it would be a weird thing for the real `FileTree.unpack()` method to raise an `AttributeError`, so I recommend against having that be the behavior in the unit test environment.

Instead, we should be providing an actual instance of a mocked `FileTree` object which actually has an `unpack()` method (either defined as part of the object or monkey-patched into it), and that method should provide a reasonable behavior (like raising `TarballUnpackError`).  This will allow the test to better discriminate between "expected exceptions" and "unexpected exceptions".